jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,461 @@
+	private native void initCifNativeThunkData(String[] argLayouts, String retLayout, boolean newArgTypes);
+	private native long invokeNative(long functionAddress, long calloutThunk, long[] argValues);
+	
+	private static final class PrivateClassLock {
+		PrivateClassLock() {}
+	}
+	private static final Object privateClassLock = new PrivateClassLock();
+	
+	static {
+		/* Resolve the required fields (specifically their offset in the jcl constant pool of VM)
+		 * which can be shared in multiple calls or across threads given the generated macros
+		 * in the vmconstantpool.xml depend on their offsets to access the corresponding fields.
+		 * Note: the value of these fields varies with different instances.
+		 */
+		resolveRequiredFields();
+	}
+	
+	ProgrammableInvoker(Addressable downcallAddr, MethodType functionMethodType, FunctionDescriptor functionDescriptor) {
+		checkIfValidLayoutAndType(functionMethodType, functionDescriptor);
+		
+		/* As explained in the Spec of LibraryLookup, the downcall must hold a strong reference to
+		 * the native library symbol to prevent the underlying native library from being unloaded
+		 * during the native calls.
+		 * 
+		 * Note: the passed-in addressable parameter can be either LibraryLookup.Symbol or MemoryAddress.
+		 */
+		functionAddr = downcallAddr;
+		funcMethodType = functionMethodType;
+		funcDescriptor = functionDescriptor;
+		cifNativeThunkAddr = 0;
+		argTypesAddr = 0;
+		generateAdapter();
+	}
+	
+	/* Map the layouts of return type & argument types to the underlying prep_cif */
+	private void generateAdapter() {
+		Optional<MemoryLayout> returnLayout = funcDescriptor.returnLayout();
+		/* Set the void layout string intended for the underlying native code as the corresponding layout doesn't exist in the Spec */
+		String retLayoutStr = (returnLayout.isPresent()) ? returnLayout.toString() : "b0[abi/kind=VOID]"; //$NON-NLS-1$
+		
+		int argLayoutCount = argLayoutArray.length;
+		String[] argLayoutStrs = new String[argLayoutCount];
+		for (int argIndex = 0; argIndex < argLayoutCount; argIndex++) {
+			argLayoutStrs[argIndex] = argLayoutArray[argIndex].toString();
+		}
+		
+		synchronized(privateClassLock) {
+			/* If a prep_cif for a given function descriptor exists, then the corresponding return & argument layouts
+			 * were already set up for this prep_cif, in which case there is no need to check the layouts.
+			 * If not the case, check at first whether the same return & argument layouts exist in the cache
+			 * in case of duplicate memory allocation for the same layouts.
+			 * 
+			 * Note: an int method (e.g. '(int,int)int') and a method function (e.g.'(boolean,boolean)boolean') share
+			 * the same C_INT layouts (e.g. 'b32[abi/kind=INT]b32[abi/kind=INT])b32[abi/kind=INT') but the underlying C
+			 * functions might be different from each other as follows:
+			 * e.g. 'int f1(int arg1, int arg2)' and 'bool f2(bool arg1, bool arg2)'.
```

Comment:
Boolean is not documented in the Spec & no support from RI. To handle this case, we choose `C_INT` for the layout of `boolean` (32bits in size as specified in JVM Spec). However, the underlying C code are different as follows:
```
(int,int)int
int add2Ints(int intArg1, int intArg2) {
	int intSum = intArg1 + intArg2;
	return intSum;
}

(boolean,boolean)boolean
int add2BoolsWithOr(int boolArg1, int boolArg2) {
	int result = (boolArg1 || boolArg2);
	return result;
}
```
In which case we can't share the `prep_cif` as they point to different functions.

I already updated the comment in there as they are unclear/misleading.

Reply:
> Boolean is not documented in the Spec & no support from RI. 

my guess is its because Boolean is a not a C type, only exists in C++ (`bool`), and I believe this JEP is only concerned with C for the time being.

> In which case we can't share the prep_cif as they are correlated to different functions in our hashmaps.

Just so I understand, in the following case are the cifs shared

```
long func(long a1, long a2) {...}
long func2(long a1, long a2) {...}
```

what about this? (assuming 64bit - ILP64)

```
BOOLEAN func(BOOLEAN a1, BOOLEAN a2) {...} //BOOLEAN is defined as uintptr_t 
long func2(long a1, long a2) {...}
```

Reply:
> my guess is its because Boolean is a not a C type, only exists in C++ (bool), and I believe this JEP is only concerned with C for the time being.


Yes. That's what I assumed previously given `C_BOOL` was not implemented eventually in OpenJDK.

> Just so I understand, in the following case are the cifs shared
> long func(long a1, long a2) {...}
> long func2(long a1, long a2) {...}

Yes, the cifs are always shared as expected as they comes with the same argument  & return layouts.

> what about this? (assuming 64bit - ILP64)
> BOOLEAN func(BOOLEAN a1, BOOLEAN a2) {...} //BOOLEAN is defined as uintptr_t 
> long func2(long a1, long a2) {...}

The JEP only requires the size of Java carrier type to match the C layout.  Assuming `BOOLEAN` is `long` in size,  `(BOOLEAN,BOOLEAN)BOOLEAN` should correspond to `(long.class, long.class, long.class)` for the method type and `(C_LONG, C_LONG, C_LONG)` for the layout, which is the same as `long func2(long a1, long a2) {...}` in terms of layouts. 

To differentiate with these two functions with the same layouts, the key of the hashmap `cachedCifNativeThunkAddr` consist of the hashcode of the function descriptor (fundamentally the layouts) and the function address to uniquely identify the specified native function.